Migrate CI to Github Actions

raynelfss opened the pull request:
<!--
⚠️ If you do not respect this template, your pull request will be closed.
⚠️ Your pull request title should be short detailed and understandable for all.
⚠️ Also, please add a release note file using reno if the change needs to be
  documented in the release notes.
⚠️ If your pull request fixes an open issue, please link to the issue.

- [x] I have added the tests to cover my changes.
- [x] I have updated the documentation accordingly.
- [x] I have read the CONTRIBUTING document.
-->

### Summary
Fixes #12807. 
The following commits move existing workflows from Azure into Github Actions.

### Details and comments
As simple as it may seem, our CI Pipeline goes through several workflows, some of which are split between Github Actions and Azure Workflows. While Azure Workflows provided us with many extra features and an isolated quota, tests seem to run much slower than they would with Github Actions (_exact numbers to be examined_). This move will (_most likely_) ensure that workflows run faster increasing our overall commit throughput.

The moved workflows include:
- Pull Requests:
    - Docs and Lint
    - Python and partial rust tests.
- Commits:
    - On push
    - Nightly tests
    - Merge queue

#### New Pull Request workflow

The new PR workflow looks as such:
<img width="702" alt="image" src="https://github.com/user-attachments/assets/2244e504-586a-463a-982e-539fb11b73a9">
Where if both `Docs and Lint` and `Preliminary Tests` pass, then the rest of the test suite runs concurrently.
For a successful run of all workflows please refer to [this example run](https://github.com/Qiskit/qiskit/pull/13474/checks?check_run_id=39780574632) from this PR.
### Tasks
- [x] Move Lint and Docs
- [x] Move Python Tests
- [x] Organize concurrency
- [x] Add nightly tests
- [x] Merge queue tests
- [ ] ~~Run nightly matrix tests when code changes.~~

### Known issues
- [x] How to properly test merge-queue, push, nightly tests.
- [ ] ~~Main nightly tests should only run when new changes in the code are introduced (as per the previous azure workflow).~~

raynelfss committed:
Initial: Move docs and lint to gh actions

raynelfss committed:
Fix: Wrong usage of map in `lint_docs` workflow

raynelfss committed:
Fix: Use ubuntu-latest for lint

raynelfss committed:
Fix: Add step to delete `doctrees` and `buildinfo`
- Added `depth: 0` to checkout to avoid reno bugs.

raynelfss committed:
Add: First drafts of Linux tests

raynelfss committed:
Add: First draft of Windows tests

raynelfss committed:
Fix: Use Poweshell formatting for scripts in Windows

raynelfss committed:
Fix: Formatting for multiline commands in Powershell

raynelfss committed:
Fix: Remove `set -e` from all windows environments.

raynelfss committed:
Fix: Use correct powershell export patterns

raynelfss committed:
Add: First draft of new Mac tests

raynelfss committed:
Fix: Add cache step

raynelfss committed:
Add: CI main workflow.

raynelfss committed:
Fix: wrong usage of workflow calls.
- Rename `tests-linux.yml` to  `test-linux.yml`.

raynelfss committed:
Fix: Rename tests to their respective os

raynelfss committed:
Fix: Add correct library path for rust tests

raynelfss committed:
Add: Nightly, Merge-queue, and Push workflows
- Add comments
- Rename `Qiskit CI workflow` to `Pull Request`

raynelfss committed:
Refactor: ci-workflow -> on-pull-request

raynelfss committed:
Remove: All azure workflow files

raynelfss committed:
Fix: Use runner for issue comment
- Add missing `.yml` extension in reusable workflow references.

raynelfss commented:
I have yet to test the nightly, merge-queue, and push tests. But I believe they should work without issues since they're mostly reusing the same workflows we use here for Pull Requests CI. I'm going to open this up for review :)

raynelfss committed:
Fix: Wrong name in MacOS tests

raynelfss committed:
Test: Remove main repository requirement
- For testing purposes.

raynelfss committed:
Merge remote-tracking branch 'upstream/main' into move-to-gh

raynelfss committed:
Fix: Adapt to latest changes
- Format documents

raynelfss committed:
Fix: Use matrix.os in test-linux

raynelfss committed:
Fix: Naming in test-linux

raynelfss committed:
Add condition to run nightly tests on PR for now

raynelfss committed:
Test: Run all new tests on PR

raynelfss committed:
Fix: Concurrency issue
- Fix wrong name being displayed for ubuntu tests.

raynelfss committed:
Fix: Further separate PR workflow from others

raynelfss committed:
Fix: Add condition for nightly-failure.

raynelfss committed:
Merge branch 'main' into move-to-gh

mtreinish commented on .github/workflows/lint_docs.yml:
```diff
@@ -0,0 +1,58 @@
+        run: tools/install_ubuntu_docs_dependencies.sh
+      - name: Run docs
+        run: |
+          make cformat
```
This is a lint task not a docs one. It's running clang format.

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,172 @@
+          name: image-test-failure-img-diffs
+          path: |
+            ./test/visual/mpl/graph/graph_results
+          if-no-files-found: error
```
We should have enough job parallelism on linux to be able to split these  out into separate jobs instead of serially executed tasks. This would reduce the runtime on this job which will be the critical path for the preliminary stage.

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,172 @@
+          rustup override set stable
+          source test-job/bin/activate
+          python -m pip install -U pip
+          python -m pip install -U build
```
In general it's better to bundle pip commands into a single execution so the depsolver can work with the shared dependencies.

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,172 @@
+        run: |
+          set -e
+          python -m pip install --upgrade pip setuptools wheel virtualenv
+          virtualenv test-job
```
I would drop virtualenv here and just do `python -m venv test-job`

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,172 @@
+    strategy:
+      fail-fast: false
+      matrix:
+        os: ["ubuntu-latest", "ubuntu-24.04-arm"]
```
I'm fine with this, but does this mean the preliminary stage for 3.9 is run in duplicate on arm and x86? I think that's actually fine but I want to know if that was actually what's going to happen and if that's your intent.

raynelfss replied:
Runs as duplicate as of now.

mtreinish reviewed (commented):
I've only gotten through the linux tests so far, but some preliminary comments.

raynelfss committed:
Fix: Address review comments
- Split python, rust, and image tests.
- Move cformat to `lint`.
- Enable selecting either ubuntu-arm or latest for ubuntu runs.

raynelfss committed:
Fix: Indentation for rust and image tests

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -33,19 +38,17 @@ on:
-    name: ${{ matrix.os }}-tests-Python-${{ inputs.python-version }}
-    runs-on: ${{ matrix.os }}
+    name: ubuntu-${{ inputs.system-suffix }}-tests-Python-${{ inputs.python-version }}
+    runs-on: ubuntu-${{ inputs.system-suffix }}
```
Instead of doing this as a suffix, I'd say change this to just be the image-type or something like that. It's not a guarantee we'll always be using ubuntu images and it's better to be explicit here.

raynelfss replied:
Added runner arguments for all workflows in 2be0b9b

mtreinish commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,36 @@
+    runs-on: ubuntu-latest
+    needs: [nightly-tests]
+    steps:
+      - uses: peter-evans/create-or-update-comment@v4
```
There's not a more official action for this?

raynelfss replied:
This is what was being used by other workflows that do the same. I could change it here but the other ones will need to be changed on a separate PR

raynelfss replied:
Switched to gh CLI in 0466f92

raynelfss committed:
Fix: Remove matrix from rust tests

mtreinish commented on .github/workflows/on-merge-queue.yml:
```diff
@@ -0,0 +1,49 @@
+name: Merge queue
+on:
+  # remove when proven to work
+  pull_request:
```
Should we mark this as on hold until you remove this?

raynelfss replied:
Agreed, so that I don't randomly forget :)

raynelfss committed:
Fix: Remove matrix from images tests

mtreinish reviewed (commented):
This is looking good. A couple of inline comments.

raynelfss committed:
Fix: Update MSRV to 1.79

raynelfss committed:
Use gh CLI to comment on issues.

raynelfss committed:
Merge branch 'main' into move-to-gh

raynelfss committed:
Fix: Install `Qiskit` before running linux rust tests.

raynelfss committed:
FIx: Add argument to specify runners for Windows and MacOS.
- Add missing MacOS 13 runner to workflows.

raynelfss committed:
Fix: Run all PR workflows together.

raynelfss committed:
Fix: Incorrect display of names for unit tests

raynelfss committed:
Test: Only run PR workflow

raynelfss committed:
Fix: Use `ubuntu-latest` instead of `latest`

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,95 @@
+  # a CI failure, and over 90% of these are in linting, documention or a test
+  # failure that would affect _any_ OS or Python version.  The goal in the
+  # first stage is to catch the vast majority of failures with minimal cost.
+  lint-docs:
```
We should set `timeout-minutes` for these jobs and limit them to an hour. That was our time budget on azure and something is fundamentally wrong if it takes that long for anything in these jobs and on the jobs on `push` and in the merge queue too.

raynelfss replied:
Added in 44b6757

mtreinish commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,37 @@
+  # remove when proven to work
+  # pull_request:
+  schedule:
+    - cron: "20 6 * * *"
```
How did you settle on this time?

raynelfss replied:
This was the same time used in Azure. Do you know if there's a way of making it only run if the code has changed? The same way `schedule/always` worked in Azure.

mtreinish replied:
I don't see anything in the docs for that in the trigger conditions. You could probably hack it together with runtime checks and storing the last sha1 run somewhere. But I wouldn't worry about that. I was more just curious if you picked this time or just copied it. It's reasonable to run things at 6:20 UTC, activity is typically pretty slow then.

mtreinish commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,37 @@
+          BODY: >
+            Randomized tests failed at commit ${{ github.sha }}.
+
+            _Logs_: ${{ github.server_url }}/${{ github.repository }}/actions/runs/${{ github.run_id }}.
```
Don't you need to specify the github token here?

See: https://docs.github.com/en/actions/writing-workflows/choosing-what-your-workflow-does/using-github-cli-in-workflows

raynelfss replied:
Added in e09daea

mtreinish commented on .github/workflows/on-merge-queue.yml:
```diff
@@ -0,0 +1,57 @@
+name: Merge queue
+on:
+  # remove when proven to work
+  # pull_request:
```
```suggestion
```

mtreinish commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,37 @@
+name: Nightly Main tests
+on:
+  # remove when proven to work
+  # pull_request:
```
```suggestion
```

raynelfss replied:
Removed in 0d98e72

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,95 @@
+on:
+  pull_request:
+    branches: [main, "stable/*"]
+  merge_group:
```
Is this is needed?

raynelfss replied:
I don't think so, this workflow should only run during pull-requests, the one for the merge queue is different and defined in `on-merge-queue`

raynelfss replied:
Removed in eb9cf50

mtreinish commented on .github/workflows/on-push.yml:
```diff
@@ -0,0 +1,34 @@
+name: Push
+on:
+  # Remove when proven to work
+  # pull_request:
```
```suggestion
```

raynelfss replied:
Removed in 708be7e

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,264 @@
+        type: boolean
+        default: false
+
+  merge_group:
```
Is this needed?

raynelfss replied:
Removed in eb9cf50

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,264 @@
+      - uses: actions/checkout@v4
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@1.79
+        if: inputs.python-version == '3.9'
```
Is this needed if we have the `rustup override stable` manually called below?

raynelfss replied:
This is only done when building from sdist, I believe we override to make sure our sdists work correctly with the latest stable version of rust.

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,95 @@
+    uses: ./.github/workflows/test-linux.yml
+    with:
+      python-version: ${{ matrix.python_version }}
+      test-rust: false
```
It looks like we're running the rust tests on 3.13: https://github.com/Qiskit/qiskit/actions/runs/15005241942/job/42162089199?pr=13474 which is duplicated work. We only need to run it once since the coverage is the same.

raynelfss replied:
This is due to a missing if statement in the test-rust job. Fixing...

raynelfss replied:
Fixed in eb9cf50

raynelfss committed:
Fix: Add missing gh token for issue comment

raynelfss committed:
Fix: Remove unnecessary merge-group from workflows

raynelfss committed:
Fix: Remove pull request stray comment

raynelfss committed:
Fix: Nightly tests were not able to comment on issues.
- Now only run on schedule

raynelfss committed:
Fix: Set timeout of 60 minutes for all jobs

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,94 @@
+    name: Docs and Lint
+    uses: ./.github/workflows/lint_docs.yml
+  preliminary-tests:
+    name: Preliminary Tests
```
If we've dropped the two stage approach and are going with maximum parallelism we should use consistent naming here. There is no preliminary or main. We also should just flatten the structure here so it's grouped by operating system/file. Right now the mix of the old and the new is confusing. Also none of the comments apply anymore.

raynelfss replied:
Fixed by a067d09 and 80c3268

mtreinish reviewed (commented):
I think this is pretty close. The big thing right now is the entire structure of the files is setup to document a two stage approach but because of the increased concurrency limits that no longer applies. We should rename the job definitions and remove all the stale comments. Or if you want to be conservative and keep a staged approach to start you can add back the dependency and save the flattening for a follow up PR.

raynelfss committed:
Fix: Use venv instead of virtualenv

raynelfss commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,266 @@
+            -e .
+          sudo apt-get update
+          sudo apt-get install -y graphviz pandoc
+          image_tests/bin/pip check
```
Is there a reason why we don't just do `source image_tests/bin/activate` here?

mtreinish replied:
It's just a habit I have when writing bash scripts. I don't like calling `activate` because of how it mangles `$PS1` and also might interact with other shell command scripting.

raynelfss commented on .github/workflows/on-push.yml:
```diff
@@ -0,0 +1,32 @@
+      test-rust: true
+      test-images: true
+      install-optionals: false
+      runner: ${{ matrix.runner }}
```
Do we want to test Windows here too?

mtreinish replied:
on azure we only run the linux 3.9 job because it's duplicated effort with the merge queue. The push job is just a sanity check that the merge was fine and doesn't really provide value so we should probably keep it minimal. I'd remove macOS and the arm job from this list too.

raynelfss replied:
Removed in ad8cd2e

raynelfss committed:
Fix: Remove stale comments about staged pipelines.

raynelfss committed:
Fix: Remove redundant naming

raynelfss committed:
Fix: "ubuntu-4.04-arm" to "ubuntu-24.04-arm" in "on-push"

raynelfss committed:
Fix: Remove "Preliminary Test" Name
- Use "Main Tests" name on all tests.

raynelfss committed:
Fix: Remove MacOS and ARM jobs from "on-push"

raynelfss committed:
Fix: Separate rust-tests and image-tests into their own workflows

raynelfss committed:
Fix: Naming convention for some tests in "on-pull-request"

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,97 @@
+        description: >
+          Decides whether we perform rust tests
+        type: boolean
+        default: false
```
Does this need to be an option here? Like we only run this once so it seems unecessary to make this switchable, just pick sdist or normal install and we can simplify the definition here.

raynelfss replied:
I'll go with normal here

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,97 @@
+        description: >
+          Decides whether we install optyional dependencies
+        type: boolean
+        default: false
```
We always need the optionals here because we're testing visualization. This shouldn't be an option and just default to true.

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,97 @@
+      - uses: actions/checkout@v4
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@1.79
+        if: inputs.python-version == '3.9'
```
I feel like building with the MSRV or not doesn't give us any value for the image tests we should just pick either stable or msrv and not make this an option. We should keep the python version config just so the config is generic for when we bump the MSPV after 3.9 goes EoL.

raynelfss replied:
Switched to stable

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,97 @@
+      QISKIT_SUPPRESS_PACKAGING_WARNINGS: Y
+      PIP_CACHE_DIR: ${{ github.workspace }}/.pip
+      QISKIT_TEST_CAPTURE_STREAMS: 1
+      HAVE_VISUAL_TESTS_RUN: false
```
What is this for?

raynelfss replied:
These flags are mostly no longer used, it seems. Tox will already enable `QISKIT_SUPPRESS_PACKAGING_WARNINGS` and `QISKIT_TEST_CAPTURE_STREAMS` to the same values here. The only reason to keep them, I assume, would be with anything that runs with stestr, unless we don't care about these variables anymore (one of which is mentioned in our [readme](https://github.com/Qiskit/qiskit/blob/main/CONTRIBUTING.md#stdoutstderr-and-logging-capture)).

If anything we should keep the `QISKIT_TEST_CAPTURE_STREAMS` whenever we test in Python so we can capture errors. The rest of them I'm not quite sure.

raynelfss replied:
Removed all env variables except for `QISKIT_TEST_CAPTURE_STREAMS` on main tests.

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,97 @@
+            -r requirements-dev.txt \
+            -e .
+          # Build and install both qiskit and qiskit-terra so that any optionals
+          # depending on `qiskit` will resolve correctly.
```
See my above comment I would just delete this

mtreinish commented on .github/workflows/on-merge-queue.yml:
```diff
@@ -0,0 +1,67 @@
+      python-version: "3.9"
+      runner: "ubuntu-latest"
+      install-optionals: true
+  test-images:
```
Not for this PR since we shouldn't change the pipeline makeup like this. But I'm not sure it's worth running the image tests in the merge queue. But we should revisit this after the migration is complete.

mtreinish commented on .github/workflows/rust-tests.yml:
```diff
@@ -0,0 +1,90 @@
+        description: >
+          Decides whether we install optyional dependencies
+        type: boolean
+        default: false
```
This will always be false we don't need to waste time installing unused dependencies in the rust tests.

mtreinish commented on .github/workflows/rust-tests.yml:
```diff
@@ -0,0 +1,90 @@
+        description: >
+          Decides whether we perform rust tests
+        type: boolean
+        default: false
```
This should also be false we don't need this to be selectable we're not testing python here or the python build system this is for running rust unit tests.

mtreinish commented on .github/workflows/rust-tests.yml:
```diff
@@ -0,0 +1,90 @@
+        run: |
+          set -e
+          python -m pip install --upgrade pip setuptools wheel
+          python -m venv test-job
```
Do we need to do this here? We're not testing python packaging we only install qiskit to cover some edge cases in our rust testing where python is still in the data model. But besides running `pip install -e .` we shouldn't need anything else.

mtreinish commented on .github/workflows/rust-tests.yml:
```diff
@@ -0,0 +1,90 @@
+          python -m pip install -U \
+              -c constraints.txt \
+              -r requirements.txt \
+              -r requirements-dev.txt \
```
You can just do `pip install -c constraints.txt -e .` here We don't need optionals or dev dependencies here we only need to install qiskit so something trying to import qiskit in rust code doesn't error.

raynelfss commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,122 @@
+          source test-job/bin/activate
+          python -m pip install -r requirements-optional.txt -c constraints.txt
+          python -m pip check
+        if: ${{ inputs.install-optionals }}
+      - name: Install optional non-Python dependencies
+        run: |
+          set -e
+          sudo apt-get update
+          sudo apt-get install -y graphviz
```
Do we still need these here? I feel like they can be removed.

mtreinish replied:
I think we have regular unit tests that run graphviz for visualization testing so we will need this on jobs that run with optionals enabled.

raynelfss committed:
Fix: Address review comments

raynelfss committed:
Fix: Introduce new arguments for workflows
- For Rust tests: we can specify rust versions using `rust-version` input.
- For test runs: We can specify to run msrv by using `use-msrv` input.

mtreinish commented:
There are syntax errors in the current version:

https://github.com/Qiskit/qiskit/actions/runs/15122611633
https://github.com/Qiskit/qiskit/actions/runs/15122611426

raynelfss committed:
Fix: Use dtolnay/rust-toolchain@master

raynelfss commented:
> There are syntax errors in the current version:
> 
> https://github.com/Qiskit/qiskit/actions/runs/15122611633 https://github.com/Qiskit/qiskit/actions/runs/15122611426

Fixed in 768694d

raynelfss committed:
Merge branch 'main' into move-to-gh

raynelfss committed:
Fix: Use global env for rust tests

raynelfss committed:
FIx: Perform pip install in one line

raynelfss committed:
Test: Remove rust toolchain action

raynelfss committed:
Fix: Remove rust toolchain step from rust tests

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,61 @@
+    steps:
+      - uses: actions/checkout@v4
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@stable
```
This doesn't do anything in practice but will expose another failure point. The `rust-toolchain.toml` will override the configured default version when `cargo` is called from the repo root since rustup is installed in the runner images. This will result in rust rustup being run twice here installing stable and then the pip call at line 42 that will call rustup internally to install the MSRV from the `rust-toolchain.toml`

```suggestion
```

raynelfss replied:
Removed in 8d54c24

mtreinish commented on .github/workflows/lint_docs.yml:
```diff
@@ -0,0 +1,60 @@
+        with:
+          python-version: "3.9"
+      - name: Install ubuntu docs dependencies
+        run: tools/install_ubuntu_docs_dependencies.sh
```
Do we need this for the lint job?

raynelfss replied:
Removed in b7f717b

mtreinish commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,37 @@
+# want to be enabled, then particular branches can override it to exclude
+# themselves by removing themselves from the trigger list.  For example, old
+# stable branches can remove `stable/*` from their copy of this file once they
+# reach their end-of-life.
```
Is this true? IIRC the schedule trigger only runs on the default branch.

raynelfss replied:
This was an old comment, this behavior probably works in Azure but not in GH Actions as it should only run scheduled workflows on the main branch. Removed in 4655adc.

mtreinish commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,37 @@
+    # For this to run on a branch, the `schedules` trigger  up at the top of this
+    # file needs to match on _both_ `main` and the branch itself.
+    if: ${{ github.repository_owner == 'Qiskit'}}
+    uses: ./.github/workflows/on-nightly.yml
```
Why does this need to be a separate file. I can understand on the other triggers where there is a clear divide between multiple job types. But for this it just seems like unnecessary layer because this is already defined in using the split job definitions. I would flatten this so there is only one nightly file to limit how much jumping around you need to do to understand the configs.

raynelfss replied:
Removed "nightly.yml" in favor of "on-nightly" in 4655adc, also updated that comment.

mtreinish commented on .github/workflows/nightly.yml:
```diff
@@ -0,0 +1,37 @@
+    # Nightly cron job.
+    #
+    # For this to run on a branch, the `schedules` trigger  up at the top of this
+    # file needs to match on _both_ `main` and the branch itself.
```
What does this mean?

raynelfss replied:
More old comments that were removed by 4655adc. Many comments were brought over thinking behavior would be similar in GH Actions, but that was not the case :)

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,90 @@
+      runner: ubuntu-latest
+      use-msrv: true
+  main-tests-rust:
+    name: Extra Tests
```
What does extra tests mean? These seem like all part of the main workflow. Do you mean non-python unit test jobs? I think that should be the distinction then. But I think instead of just saying extra tests here "Rust unit tests" and "Image comparison tests" for the names here would be better.

raynelfss replied:
Changed to "Rust Unit Tests" and "Image Comparison Tests" in b7f717b

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,90 @@
+    name: Docs and Lint
+    uses: ./.github/workflows/lint_docs.yml
+  main-tests-ubuntu-3-9:
+    name: Main Tests
```
Maybe "Python unit tests" instead?

raynelfss replied:
Changed in b7f717b

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,125 @@
+      - uses: actions/checkout@v4
+      - name: Install Rust toolchain
+        uses: dtolnay/rust-toolchain@1.79
+        if: ${{ inputs.use-msrv }}
```
This is actually just likely to get out of sync, if `use-msrv` is set to true you're potentially installing rust twice if install from sdist is used and not actually running with a different rust versions. You would need to also set the appropriate override when calling cargo (via pip) if you wanted `false` case to be respected fully.

I would just remove this flag and option from the test jobs. This is different from the azure config as that always used MSRV and never ran with Rust stable (it did over a year ago but we moved away from that). We can discuss doing this in the future, but we shouldn't make changes like this as part of a migration PR.

raynelfss replied:
Removed in 8d54c24

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,125 @@
+        run: |
+          set -e
+          # Use stable Rust, rather than MSRV, to spot-check that stable builds properly.
+          rustup override set stable
```
If we were doing this (which we shouldn't in this PR) wouldn't this need to be conditioned on `use-msrv` not install from sdist?

raynelfss replied:
Removed in 8d54c24

mtreinish commented on .github/workflows/test-mac.yml:
```diff
@@ -0,0 +1,99 @@
+          Decides whether we use MSRV or stable for rust toolchain.
+        required: false
+        type: boolean
+        default: false
```
Same comment from linux remove this option and always use msrv.

raynelfss replied:
Removed in 8d54c24

mtreinish commented on .github/workflows/test-windows.yml:
```diff
@@ -0,0 +1,100 @@
+          Decides whether we use MSRV or stable for rust toolchain.
+        required: false
+        type: boolean
+        default: false
```
Same here remove this option and always use msrv.

raynelfss replied:
Removed in 8d54c24

raynelfss committed:
Fix: Clean up nightly matrix tests.
- Consolidate nightlty jobs into one file.
- Remove strange comment

raynelfss committed:
Fix: Remove `dtolnay/rust-toolchain` from tests workflows.
- Remove rust override from sdist installation.

raynelfss committed:
Fix: Remove ubuntu docs dependencies installation in lint workflow
- Rename main tests to "Python Unit Tests", and other tests to "Rust Unit Tests" and "Image Comparison Tests".

raynelfss committed:
Fix: Install tox dependency in lint workflow

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,59 @@
+        with:
+          name: image-test-failure-img-diffs
+          path: |
+            ./test/visual/mpl/graph/graph_results
```
This is missing results, it needs to include the circuit results directory too. There are two sets of tests that write out the results to different paths.

raynelfss replied:
Fixed in c1f1a8a

mtreinish commented on .github/workflows/lint_docs.yml:
```diff
@@ -0,0 +1,60 @@
+        run: python -m pip install --upgrade tox
+      - name: Run lint
+        run: |
+          make cformat
```
For those following along in azure we ran a script `tools/install_ubuntu_c_dependencies.sh` that installs clang format which is missing here. But it doesn't matter because the gha runner image already includes clang format: https://github.com/actions/runner-images/blob/main/images/ubuntu/Ubuntu2404-Readme.md#language-and-runtime which is the same version. We can probably delete that bash script but lets save that for a followup.

mtreinish commented on .github/workflows/test-mac.yml:
```diff
@@ -0,0 +1,88 @@
+        uses: actions/setup-python@v5
+        with:
+          python-version: ${{ inputs.python-version }}
+          architecture: arm64
```
Is this correct? Shouldn't we only use the arm64 version of python on the macOS-14 runner?

raynelfss replied:
You are correct! Do I make it do so conditionally? Like e.g.
```yaml
          # This may not wokr
          architecture: ${{ inputs.runner == "macos-14" && "arm64" ||  "x64" }}
```

mtreinish replied:
Does it work out of the box if you just remove this flag? Like I feel like the action should just use the native architecture if you don't specify anything.

raynelfss replied:
I don't see any reason why it shouldn't work. It defaults to the architecture of the runner as per [documentation](https://github.com/actions/setup-python/tree/v5/?tab=readme-ov-file#supported-architectures).

raynelfss replied:
Fixed in 4dc9eae

mtreinish commented on .github/workflows/test-mac.yml:
```diff
@@ -0,0 +1,88 @@
+            -r requirements-dev.txt \
+            -e .
+            # Build and install both qiskit and qiskit-terra so that any optionals
+            # depending on `qiskit` will resolve correctly.
```
```suggestion
```

raynelfss replied:
Removed in 9ee9683

mtreinish commented on .github/workflows/test-windows.yml:
```diff
@@ -0,0 +1,89 @@
+            -r requirements-dev.txt `
+            -e .
+          # Build and install both qiskit and qiskit-terra so that any optionals
+          # depending on `qiskit` will resolve correctly.
```
```suggestion
```

raynelfss replied:
Removed in 9ee9683

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,114 @@
+            -r requirements-dev.txt \
+            -e .
+          # Build and install both qiskit and qiskit-terra so that any optionals
+          # depending on `qiskit` will resolve correctly.
```
```suggestion
```

raynelfss replied:
Removed in 9ee9683

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,114 @@
+            -r requirements-dev.txt \
+            dist/qiskit-*.tar.gz
+          # Build and install both qiskit and qiskit-terra so that any optionals
+          # depending on `qiskit` will resolve correctly.
```
```suggestion
```

raynelfss replied:
Removed in 9ee9683

raynelfss committed:
Fix: Add missing circuit results

raynelfss committed:
Chore: Remove stale comment

raynelfss committed:
Fix: Remove architecture argument from mac tests

mtreinish commented on .github/workflows/on-merge-queue.yml:
```diff
@@ -0,0 +1,65 @@
+on:
+  merge_group:
+    branches: [main, "stable/*"]
+concurrency:
+  group: ${{ github.repository }}-${{ github.ref }}-${{ github.head_ref }}-${{ github.workflow }}
+  cancel-in-progress: true
```
This doesn't hurt, but it really shouldn't do anything because we should ever have 2 groups at the same time for commits that have entered the merge queue. If someone pushes a new commit to a enqueued PR it wouldn't be mergeable anymore and a new merge queue job wouldn't start. Lets just leave it though because it doesn't hurt to have this protection in place here too.

mtreinish commented on .github/workflows/on-merge-queue.yml:
```diff
@@ -0,0 +1,65 @@
+name: Merge queue
+on:
+  merge_group:
+    branches: [main, "stable/*"]
```
I'm wondering what's going to happen after we merge this. We probably aren't going to backport this PR and we'll still use azure on stable branches until they age out (`stable/2.0` goes EoL in about a month). We might want to backport it to `stable/1.4` since that's supported longer. I'm wonder if we will need to remove this until we are using gha on all the supported stable branches.

mtreinish commented on .github/workflows/on-merge-queue.yml:
```diff
@@ -0,0 +1,65 @@
+    uses: ./.github/workflows/test-windows.yml
+    strategy:
+      matrix:
+        runner: ["windows-latest"]
```
Using a matrix here seems a bit unnecessary since there isn't more than 1 configuration but it's fine for now. I can see a world where we add windows arm support in several years so lets leave the flexibility for the future.

raynelfss replied:
Removed in 75eb49c

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,88 @@
+---
+# Full PR suite.  PRs need to pass this pipeline in order to be moved to the
+# merge queue, where they'll use the next rule as the branch-protection rule
+# for the final merge to the base branch.
```
```suggestion
# Full PR suite. PRs need to pass this pipeline in order to be moved to the
# merge queue, where they'll use the on-merge-queue workflow for the final
# merge to the base branch.
```

raynelfss replied:
Fixed in 3a00da3

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,88 @@
+    uses: ./.github/workflows/test-linux.yml
+    with:
+      python-version: "3.9"
+      install-from-sdist: true
```
Should this be false to match the other 3.9 job?

raynelfss replied:
Fixed in 3a00da3, merged into a matrix job

mtreinish commented on .github/workflows/on-pull-request.yml:
```diff
@@ -0,0 +1,88 @@
+    with:
+      python-version: "3.9"
+      install-from-sdist: true
+      install-optionals: false
```
Shouldn't this be true to match the other 3.9 job? Or do the optional packages not all work on aarch64? (I vaguely remember that might have been the case).

raynelfss replied:
Fixed in 3a00da3, merged into a matrix job

mtreinish commented on .github/workflows/on-push.yml:
```diff
@@ -0,0 +1,28 @@
+---
+# Push to main or the stable branches.  The triggering branches also need to
+# be in the triggers at the top of this file.
```
This is the top of the file...
```suggestion
# Run on push to main or the stable branches.
```

mtreinish commented on .github/workflows/rust-tests.yml:
```diff
@@ -0,0 +1,45 @@
+      QISKIT_SUPPRESS_PACKAGING_WARNINGS: Y
+      PIP_CACHE_DIR: ${{ github.workspace }}/.pip
+      QISKIT_TEST_CAPTURE_STREAMS: 1
+      HAVE_VISUAL_TESTS_RUN: false
```
None of these will have any effect here
```suggestion
```

raynelfss replied:
Removed in fb635da

mtreinish commented on .github/workflows/rust-tests.yml:
```diff
@@ -0,0 +1,45 @@
+          python -m pip install -c constraints.txt -e .
+      - name: Run Rust Tests
+        # Note that we use the virtual env with Qiskit installed to run the Rust
+        # tests since some of them still depend on Qiskit's Python API via PyO3.
```
There is no virtual environment used in this job.
```suggestion
```

raynelfss replied:
Removed in fb635da

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,110 @@
+          export PYTHONHASHSEED=$(python -S -c "import random; print(random.randint(1, 4294967295))")
+          echo "PYTHONHASHSEED=$PYTHONHASHSEED"
+          stestr run
+          popd
```
This is missing a filter step post-run that is used to the filter the stestr history. This is what: https://github.com/Qiskit/qiskit/blob/main/.azure/test-linux.yml#L135-L137 does. We'll need this on all platform jobs if we're caching the stestr run times for better scheduling.

Without this you'll have continual growth of the stestr history because the run history will be incremented on each run growing to use all the cache space and being very slow to upload and download.

raynelfss replied:
Would we also need https://github.com/Qiskit/qiskit/blob/22367fca2dfa7cc22d5b4167b74279697f838f96/.azure/test-linux.yml#L138-L140 (except line 139) here too?

mtreinish replied:
Yes, you will need the copy so the cache action has the stestr timing data to upload. This copy is only needed when you're not running in the repo like this job is. In the other OSes you don't need the copy because .stestr in the repo root is where you have the cache action pointing to.

mtreinish replied:
Oh actually that raises a question why are we running the tests in `/tmp/terra-tests` unconditionally. I would expect this to only be for the sdist job where we want to test the install from a package installation and avoid issues with the CWD getting imported by mistake.

raynelfss replied:
Tried to fix in 1e0466d

raynelfss committed:
Chore: Remove unused matrix from "on-merge-queue"

raynelfss committed:
Chore: Remove unused environment variables from "rust-tests".
- Remove stale comment.

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Fix: Streamline pull request resting
- Use matrix for 3.9 tests
- Remove stale comment.

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Fix: Add missing stestr history filter step

raynelfss committed:
Fix: Add missing copy and publish images step for "test-linux"

raynelfss committed:
Update .github/workflows/on-push.yml

Co-authored-by: Matthew Treinish <mtreinish@kortar.org>

raynelfss committed:
Fix: Use perl instead of grep for mac
- Remove `set -e` from Windows.

raynelfss committed:
Fix: Add different name to each copied images name.

raynelfss committed:
Fix: Errors in new stestr filtering step
- Remove redundant popd after test runs on runs from source.
- Remove cp step from filtering stestr step.

raynelfss committed:
Fix: Rervert command to the one used in old workflow for mac.

mtreinish commented on .github/workflows/image-tests.yml:
```diff
@@ -0,0 +1,60 @@
+          name: image-test-failure-img-diffs
+          path: |
+            ./test/visual/mpl/graph/graph_results
+            ./test/visual/mpl/circuit/circuit_results
```
This is also missing: `test/visual/mpl/visual_test_failures`

raynelfss committed:
Fix: Add missing path in image tests.

mtreinish commented on .github/workflows/test-linux.yml:
```diff
@@ -0,0 +1,146 @@
+    steps:
+      - uses: actions/checkout@v4
+      - name: Set up Python ${{ inputs.python-version }}
+        id: python-outputs
```
What is this used for?

raynelfss replied:
Nothing :) Removed in 50b4883

raynelfss committed:
Chore: Remove usage of `actions/setup-pythoh.id`

mtreinish commented on .github/workflows/test-mac.yml:
```diff
@@ -0,0 +1,91 @@
+        run: |
+          set -e
+          source test-job/bin/activate
+          pip install -r requirements-optional.txt -c constraints.txt
```
Should you `pip check` here instead of above?

raynelfss replied:
Added to the line after in cf2b8d3

mtreinish commented on .github/workflows/test-windows.yml:
```diff
@@ -0,0 +1,92 @@
+          python -m venv test-job
+          .\test-job\Scripts\activate
+          python -m pip install -U pip setuptools wheel
+          python -m pip install -U `
```
I guess \` is the right syntax on windows cmd/powershell for the bash equivalent of `\`? I didn't realize that

raynelfss replied:
Yeah, apparently this is the powershell syntax for continued line [(source).](https://devblogs.microsoft.com/scripting/powertip-line-continuation-in-powershell/)

mtreinish commented on .github/workflows/test-windows.yml:
```diff
@@ -0,0 +1,92 @@
+            -r requirements.txt `
+            -r requirements-dev.txt `
+            -e .
+          pip check
```
This probably isn't needed here because it's a single pip command. This was originally needed before pip added a new dependency solver to validate your weren't relying on implicit behavior that would break. It's good to run on two successive pip calls still to check that you didn't break a dependency from the first install with the second (because the depsolver is only invoked per session and not shared across multiple calls).

raynelfss committed:
Fix: Add missing `pip check` on "test_mac"

mtreinish reviewed (dismissed):
I think this LGTM now, lets wait for the test PR to pass the jobs to make sure and then we can make the migration.

raynelfss committed:
Fix: "on-merge" names now match "on-pull-request" names
